numpyro/primitives.py

```diff
@@ -265,19 +265,16 @@ def process_message(self, msg):
         frame = CondIndepStackFrame(self.name, self.dim, self.subsample_size)
         cond_indep_stack.append(frame)
         expected_shape = self._get_batch_shape(cond_indep_stack)
-        dist_batch_shape = msg['fn'].batch_shape if msg['type'] == 'sample' else ()
-        overlap_idx = max(len(expected_shape) - len(dist_batch_shape), 0)
-        trailing_shape = expected_shape[overlap_idx:]
-        # e.g. distribution with batch shape (1, 5) cannot be broadcast to (5, 5)
-        broadcast_shape = lax.broadcast_shapes(trailing_shape, dist_batch_shape)
-        if broadcast_shape != dist_batch_shape:
-            raise ValueError('Distribution batch shape = {} cannot be broadcast up to {}. '
-                             'Consider using unbatched distributions.'
-                             .format(dist_batch_shape, broadcast_shape))
-        batch_shape = expected_shape[:overlap_idx]
-        if 'sample_shape' in msg['kwargs']:
-            batch_shape = lax.broadcast_shapes(msg['kwargs']['sample_shape'], batch_shape)
-        msg['kwargs']['sample_shape'] = batch_shape
+        if msg['type'] == 'sample':
+            dist_batch_shape = msg['fn'].batch_shape
+            if 'sample_shape' in msg['kwargs']:
+                dist_batch_shape = msg['kwargs']['sample_shape'] + dist_batch_shape
+                msg['kwargs']['sample_shape'] = ()
+            overlap_idx = max(len(expected_shape) - len(dist_batch_shape), 0)
+            trailing_shape = expected_shape[overlap_idx:]
+            broadcast_shape = lax.broadcast_shapes(trailing_shape, dist_batch_shape)
+            batch_shape = expected_shape[:overlap_idx] + broadcast_shape
+            msg['fn'] = msg['fn'].expand(batch_shape)
```

Comment:
Should we also expand when sample_shape is available at process or postprocess trace handler for consistency?

Reply:
Also, maybe at the constructor of ExpandedDist, we can check if base dist is expanded dist. If so, we just need to store `base_dist.base_dist`. This is convenient for inspecting the sites under multiple plate statements.

Reply:
> Should we also expand when sample_shape is available at process or postprocess trace handler for consistency?

`sample_shape` should already incorporated. Right now, the behavior is that the distribution's intrinsic shape is `sample_shape + dist.batch_shape` which needs to broadcast with all the plates. This was what we had earlier, so the behavior is kept consistent. We can instead do `sample_shape + broadcasted shape from plates and batch shape`, but that can be surprising. Ideally, we shouldn't use `sample_shape` with `plate` to avoid this kind of confusion.

> Also, maybe at the constructor of ExpandedDist, we can check if base dist is expanded dist. If so, we just need to store base_dist.base_dist. 

Let me see how we are doing this in Pyro, I agree multiple layers of `ExpandedDist` is not great for debugging.

Reply:
I see. It is a bit confusing. I like your change because it solves the shape broadcasting issue that we have before (the ValueError in the previous code).

Reply:
Regarding..
> Also, maybe at the constructor of ExpandedDist, we can check if base dist is expanded dist. If so, we just need to store base_dist.base_dist.

Just by using nested plates, we will never hit this condition where the `base_dist` is already an `ExpandedDistribution`. This can only happen if the user uses `.expand` and `plate` together. Does that make sense? We can still introduce this optimization, but it shouldn't affect the vast majority of models.

Reply:
I see. I already add this optimization in #617 (just to make sure that we can use `TransformReparam` with multiple `.expand` op).